frontend/src/map/GameMap.js

```diff
@@ -1,14 +1,71 @@
+function tileKey(x, y, z) {
+    return [x, y, z].join(',');
+}
+
+function cellKey(x, y) {
+    return [x, y].join(',');
+}
+
 class GameMap {
-    constructor(width, depth, height) {
-        this.width = width
-        this.depth = depth
-        this.height = height
+    constructor() {
+        // For simplicity, assume for now that the minimum value for each
+        // dimension is 0.
+        this.maxX = 0;
+        this.maxY = 0;
+        this.maxZ = 0;
+
+        this.mapData = new Map();
+        this.elevationData = new Map();
+    }
+
+    get width() {
+        return this.maxX;
+    }
+
+    get depth() {
+        return this.maxY;
+    }
+
+    get height() {
+        return this.maxZ;
+    }
+
+    forEachTile(callbackFn) {
+        let results = []
+        for (let [key, tile] of this.mapData) {
+            const coordinates = key.split(',').map((x) => parseInt(x, 10));
+            results.push(callbackFn(tile, coordinates));
+        }
+        return results
+    }
+
+    getTile(x, y, z) {
+        return this.mapData.get(tileKey(x, y, z)) || undefined;
+    }
+
+    isOccupied(x, y, z) {
+        return this.getTile(x, y, z) != undefined;
+    }
+
+    getElevation(x, y) {
+        return this.elevationData.get(cellKey(x, y)) || 0;
+    }
+
+    addTile(x, y, z, tile) {
+        this.mapData.set(tileKey(x, y, z), tile);
+
+        const cell = cellKey(x, y);
+        this.elevationData.set(cell, Math.max(this.elevationData.get(cell) ?? 0, z));
 
-        this.mapData = new Array()
+        this.maxX = Math.max(this.maxX, x + 1);
```

Comment:
Just trying to make sure I understand. All maps start out 0x0x0, but as you add tiles, the map grows to fit them. So you're trying to allow tiles to be built vertically, but also horizontally, changing the size of the map.

Reply:
Exactly. Since we're going to allow terraforming (adding & removing tiles), I wanted the `GameMap` class to manage the size itself instead of relying on the user to update it correctly.

Reply:
Theoretically, horizontally would be possible but if we should actually include that in the gameplay is another question.